servicetalk-transport-netty-internal/src/main/java/io/servicetalk/transport/netty/internal/DefaultNettyConnection.java

```diff
@@ -938,6 +939,10 @@ public void userEventTriggered(ChannelHandlerContext ctx, Object evt) {
                         DefaultNettyConnection.class, "userEventTriggered(ChannelOutputShutdownEvent)"));
             } else if (evt == SslCloseCompletionEvent.SUCCESS) {
                 connection.closeHandler.channelCloseNotify(ctx);
+            } else if (evt == ChannelInputShutdownEvent.INSTANCE) {
+                // We've received a hangup but there may be more bytes buffered by the socket so let's read them into
+                // our channel buffer. We might not receive the ChannelInputShutdownReadComplete event until we try.
+                ctx.read();
```

Comment:
While I don't mind merging this change, those are my main concerns with netty:
1. This is an unexpected behavior change in Netty 4.2. There could be other libraries that look only for `ChannelInputShutdownReadComplete`. As we could see here and in #3013, debugging this problem is not trivial, requires deep netty knowledge, and pedantic test coverage to catch it. 
2. While it doesn't affect HTTP/2 stream channels for any of our tests (yet), we should revert https://github.com/netty/netty/pull/13395. Otherwise, this new path won't execute for HTTP/2 stream channel.

Have a look at https://github.com/netty/netty/pull/13395 description. Based on that, it feels like there is high probability other libraries that work with Netty 4.1 also look only for `ChannelInputShutdownReadComplete` event but not for `ChannelInputShutdownEvent`. Therefore, (1) is likely to happen.

Reply:
I tend to agree that there is a good probability other netty users dependent on this behavior despite how weird it may be. It's also not yet clear if the behavior is consistent across all io implementations.

Another thing to consider for ST is that if kept a read outstanding for otherwise idle client channels we wouldn't suffer from the behavior change in the epoll implementation _and_ we could detect hangups with the NIO implementation as well.

Reply:
Yeah I agree with @idelpivnitskiy here after thinking a bit more about it. I guess best is to just pull in the netty pr.